File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/data/input/GoogleSheetsDataLoaderImpl.java
package uk.co.cerihughes.mgm.data.input;

import org.apache.http.HttpEntity;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.ResponseHandler;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;

import java.io.IOException;

public class GoogleSheetsDataLoaderImpl implements GoogleSheetsDataLoader {
    private static final String url = "https://spreadsheets.google.com/feeds/list/1AeTjG2coPkJ_3XSsSUWmu1Pb5Tpm9YdHP3M3iBtgUew/od6/public/values?alt=json";

    public String loadJsonData() throws IOException {
        HttpClient httpclient = new DefaultHttpClient();
        HttpGet httpget = new HttpGet(url);

        ResponseHandler<String> responseHandler = response -> {
            int status = response.getStatusLine().getStatusCode();
            if (status >= 200 && status < 300) {
                HttpEntity entity = response.getEntity();
                return entity != null ? EntityUtils.toString(entity) : null;
            } else {
                throw new ClientProtocolException("Unexpected response status: " + status);
            }
        };
        return httpclient.execute(httpget, responseHandler);
    }
}

File: mgm-gcp/src/main/java/uk/co/cerihughes/mgm/data/input/GoogleSheetsDataLoaderImpl.kt
package uk.co.cerihughes.mgm.data.input

import com.google.api.client.http.HttpExecuteInterceptor
import com.google.api.client.http.HttpRequest
import com.google.api.client.http.HttpRequestInitializer
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.JsonFactory
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.sheets.v4.Sheets
import uk.co.cerihughes.mgm.Secrets
import java.io.IOException

class GoogleSheetsDataLoaderImpl : GoogleSheetsDataLoader {
    companion object {
        private val jsonFactory: JsonFactory = GsonFactory.getDefaultInstance()
    }

    private fun getSheets(): Sheets {
        val transport = NetHttpTransport.Builder().build()
        val httpRequestInitializer = HttpRequestInitializer { request: HttpRequest ->
            request.interceptor = HttpExecuteInterceptor { intercepted: HttpRequest ->
                intercepted.url["key"] = Secrets.googleSheetsApiKey
            }
        }
        return Sheets.Builder(transport, jsonFactory, httpRequestInitializer)
            .build()
    }

    private fun getValues(): List<List<Any?>?>? {
        try {
            return getSheets()
                .spreadsheets()
                .values()[Secrets.googleSheetsId, "Sheet1!A2:H"]
                .execute()
                .getValues()
        } catch (ex: IOException) {
            return null
        }
    }

    override fun loadSheetsData(): List<List<String>>? {
        val outerList = getValues() ?: emptyList<List<String>>()
        return outerList
            .mapNotNull { innerList ->
                innerList?.mapNotNull { it?.toString() }
            }
    }
}
